File: src/main/java/sage/web/TagController.java
package sage.web;

import java.util.Collection;

import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.web.bind.annotation.PathVariable;
import org.springframework.web.bind.annotation.RequestMapping;
import org.springframework.web.bind.annotation.RequestParam;
import org.springframework.web.bind.annotation.RestController;
import sage.entity.Tag;
import sage.service.TagChangeService;
import sage.service.TagService;
import sage.transfer.TagCard;
import sage.transfer.TagChangeRequestCard;
import sage.transfer.TagNode;
import sage.web.auth.Auth;

@RestController
@RequestMapping("/tag")
public class TagController {
  @Autowired
  private TagService tagService;
  @Autowired
  private TagChangeService tagChangeService;

  @RequestMapping("/card/{id}")
  public TagCard tagCard(@PathVariable Long id) {
    return tagService.getTagCard(id);
  }

  @RequestMapping("/tree")
  public TagNode tagTree() {
    return tagService.getTagTree();
  }

  @RequestMapping("/new")
  public long create(@RequestParam String name,
                     @RequestParam(required = false) Long parentId,
                     @RequestParam(required = false) String intro) {
    Auth.checkCuid();
    if (parentId == null) parentId = Tag.Companion.getROOT_ID();
    return tagChangeService.newTag(name, parentId, intro);
  }

  @RequestMapping("/{id}/move")
  public void move(@PathVariable Long id, @RequestParam Long parentId) {
    tagChangeService.requestMove(Auth.checkCuid(), id, parentId);
  }

  @RequestMapping("/{id}/rename")
  public void rename(@PathVariable Long id, @RequestParam String name) {
    tagChangeService.requestRename(Auth.checkCuid(), id, name);
  }

  @RequestMapping("/{id}/setIntro")
  public void setIntro(@PathVariable Long id, @RequestParam String intro) {
    tagChangeService.requestSetIntro(Auth.checkCuid(), id, intro);
  }

  @RequestMapping("/{id}/requests")
  public Collection<TagChangeRequestCard> requests(@PathVariable Long id) {
    Auth.checkCuid();
    return tagChangeService.getRequestsOfTag(id);
  }

  @RequestMapping("/{id}/scope-requests")
  public Collection<TagChangeRequestCard> scopeRequests(@PathVariable Long id) {
    Auth.checkCuid();
    return tagChangeService.getRequestsOfTagScope(id);
  }
}

File: src/main/kotlin/sage/web/page/TagController.kt
package sage.web.page

import org.springframework.stereotype.Controller
import org.springframework.web.bind.annotation.*
import org.springframework.web.servlet.ModelAndView
import sage.entity.Blog
import sage.entity.Tag
import sage.transfer.BlogPreview
import sage.util.Strings
import sage.web.auth.Auth
import sage.web.context.BaseController

@Controller
@RequestMapping("/tags")
open class TagController : BaseController() {

  @RequestMapping("/new", method = arrayOf(RequestMethod.POST))
  @ResponseBody
  open fun create(@RequestParam name: String,
                  @RequestParam(required = false) parentId: Long?,
                  @RequestParam(defaultValue = "false") isCore: Boolean,
                  @RequestParam(defaultValue = "") intro: String): String {
    val uid = Auth.checkUid()
    val tag = tagService.create(uid, name, parentId ?: Tag.ROOT_ID, isCore, intro)
    return "/tags/${tag.id}"
  }

  @RequestMapping("/{id}")
  open fun get(@PathVariable id: Long): ModelAndView {
    val tag = Tag.get(id)
    val blogs = Blog.where().`in`("tags.id", tag.getQueryTags().map { it.id }).findList()
        .sortedByDescending { it.whenCreated }.map(::BlogPreview)

    val (coreTags, nonCoreTags) = tag.children.map { it.toTagLabel() }.partition { it.isCore }
    val relatedTags = null
    val sameNameTags = tagService.getSameNameTags(id).map { it.toTagLabel() }

    frontMap().attr("id", id)

    return ModelAndView("tag-page")
        .addObject("tag", tag)
        .addObject("blogs", blogs)
        .addObject("coreTags", coreTags)
        .addObject("nonCoreTags", nonCoreTags)
        .addObject("relatedTags", relatedTags)
        .addObject("sameNameTags", sameNameTags)
        .addObject("countPendingRequestsOfTagScope", tagChangeService.countPendingRequestsOfTagScope(id))
        .addObject("countPendingRequestsOfTag", tagChangeService.countPendingRequestsOfTag(id))
  }

  @RequestMapping("/{id}/rss")
  fun rss(@PathVariable id: Long): ModelAndView {
    val blogs = Blog.where().`in`("tags", Tag.ref(id)).findList()
    response.contentType = "text/xml"
    blogs.forEach {
      it.content = Strings.escapeXmlInvalidChar(it.content)
    }
    return ModelAndView("rss").addObject("blogs", blogs).addObject("name", Tag.get(id).name)
  }
}